### ISSUE TITLE
Make AOT IR less dynamically typed

### ISSUE THREAD

**Issue by ltratt** [ykjit/yk#1088]
Currently the AOT IR (in `aot_ir.rs`) is heavily, and not entirely intuitively, dynamically typed. Although there are several things that we could probably improve here, the main problem is that `Instruction` is a `struct` that references both (integer) `OpCode` and the (enum) `Operand`: this leads to code having to know which operand in a list pertains to which expected thing and so on.

We can take a leaf out of the JIT IR (`jit_ir.rs`) and change this to (roughly):

```rust
enum Instruction {
  PtrAdd{lhs: Operand, rhs: Operand},
  Load{src: ..., target: Operand},
  ...
}

enum Operand {
    Constant(ConstantIdx),
    ValueResultingFromAnInstruction(InstructionIdx)
}
```

In other words, `Operand` slims down, while `Instruction` becomes an enum with all the instruction variants. This will get rid of most of the dynamically typed aspects and also, I'm fairly sure, save us both time and memory.

**Comment by vext01** [comment]
I believe this has been fixed a while back.

`Inst` is now an enum with different innards depending upon discriminant (opcode):
https://github.com/ykjit/yk/blob/ef19fce841dcf695764bfdd25993e7a5f2ad39dd/ykrt/src/compile/jitc_yk/aot_ir.rs#L782
